### PR TITLE
[DSV3] Add PP support for DSV3

### PR DESCRIPTION
**Pull request by H-Huang**
This takes a different approach from what we do in LLaMa3/4:
1. Dynamically replaces the `forward()` method only when using PP (therefore no `model.py` changes required)
2. New helper method to explicitly specify the modules to include

Can run the 16B DSV3 on 8 GPUs with PP:
```
NGPU=8 LOG_RANK=0,7 CONFIG_FILE="./torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml" ./run_train.sh --parallelism.pipeline_parallel_degree 8 --parallelism.pipeline_parallel_schedule Interleaved1F1B
```

TODO:
- upstream `pipeline_module_split` to `torch.distributed.pipelining`?

### CODE REVIEW

File: torchtitan/models/deepseek_v3/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+Download tokenizer:
```
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/README.md]: Thanks for making this!

File: torchtitan/models/deepseek_v3/infra/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    This method is only used when pipeline parallelism is enabled.
+    If model attributes are None, they are skipped in the forward pass.
+
+    Args:
+        tokens (torch.Tensor): Input tensor of token IDs with shape (batch_size, seq_len).
+
+    Returns:
+        torch.Tensor: Logits tensor of shape (batch_size, vocab_size).
+    """
+    h = self.tok_embeddings(tokens) if self.tok_embeddings is not None else tokens
```
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/infra/pipeline.py]: Does these "if xx is not None else xxx" also applies to the forward function if there's no PP, should we add it to forward directly?

Comment (H-Huang) [reply torchtitan/models/deepseek_v3/infra/pipeline.py]: good question, it only applies to PP. Since in PP we are splitting the model and then setting some modules to `None`. For llama, we directly modified the model in titan to make this work:

https://github.com/pytorch/torchtitan/blob/aefe15a32282885bc2a6ade3d373b23d9ebf759d/torchtitan/models/llama3/model/model.py#L419-L432

compared to regular llama:

https://github.com/meta-llama/llama3/blob/a0940f9cf7065d45bb6675660f80d305c041a754/llama/model.py#L285-L301

But I thought it would be cleaner to leave the model forward as-is and only modify if using PP. Maybe that will come back to bite me lol

File: torchtitan/train.py [hunk 3]
```diff
@@ -551,5 +550,6 @@ def close(self) -> None:
             trainer.close()
 
         if torch.distributed.is_initialized():
-            torch.distributed.destroy_process_group()
+            # TODO: why does destroy_process_group() hang when another process is stuck?
```
Comment (tianyu-l) [inline torchtitan/train.py]: see https://github.com/pytorch/torchtitan/pull/1342

File: torchtitan/models/deepseek_v3/infra/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+from torchtitan.config_manager import JobConfig
+from torchtitan.distributed import ParallelDims
+from torchtitan.distributed.pipeline import (
+    build_pipeline_schedule,
+    stage_ids_this_rank,
+)
+from torchtitan.protocols.train_spec import DeviceType, ParallelizeFunction
+from torchtitan.tools.logging import logger
+
+# Should I use BaseModelArgs instead of DeepSeekV3ModelArgs?
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/infra/pipeline.py]: `DeepSeekV3ModelArgs` sounds fine as it is only used in `pipeline_deepseekv3`?
also it access `model_config.n_layers` which is not in `BaseModelArgs`

File: torchtitan/models/deepseek_v3/infra/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+    device: DeviceType,
+    model_config: DeepSeekV3ModelArgs,
+    parallelize_fn: ParallelizeFunction,
+    loss_fn: LossFunction,
+) -> tuple[_PipelineSchedule, list[nn.Module], bool, bool]:
+    pp_mesh = world_mesh["pp"]
+    
+    # Determine the number of virtual stages based on schedule type
+    schedule_class = get_schedule_class(job_config.parallelism.pipeline_parallel_schedule)
+    is_single_stage_schedule = schedule_class.__name__ in ["PipelineScheduleSingle"]
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/infra/pipeline.py]: in Llama 3 we seem to use `is_single_stage_schedule = issubclass(schedule_class, PipelineScheduleSingle)`

File: torchtitan/models/deepseek_v3/infra/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    # Ensure each stage gets at least the weight of input/output modules
+    if layers_per_stage < max(input_weight, output_weight):
+        raise ValueError(
+            f"Layers per stage ({layers_per_stage}) must be >= max(input_weight={input_weight}, output_weight={output_weight})"
+        )
+    
+    module_names_per_stage = []
+    current_layer = 0
+    
+    for stage_idx in range(num_stages):
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/infra/pipeline.py:83]: It seems the LoC is quite longer than https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/pipeline.py#L106-L121

as you are not generating the split points but generating actual module names per stage. (I'm assuming function-wise they are the same but I didn't check.)
But later on the complexity in `pipeline_deepseekv3_module_split` doesn't seem to be saved.

I'd like to learn more about the reasonings behind the change of UI.

File: torchtitan/train.py (hunk 3 above)
Comment (H-Huang) [reply torchtitan/train.py]: ah nice, let me try to rebase and see if that fixes things!

File: torchtitan/models/deepseek_v3/infra/pipeline.py (hunk 6 above)
Comment (H-Huang) [reply torchtitan/models/deepseek_v3/infra/pipeline.py:83]: Module names are more flexible since they can be applied for all models. `generate_module_names_per_stage` is specific to deepseek-v3; however, the new helper method i refactored `pipeline_module_split` is model agnostic. So I am thinking of upstreaming this as a utility in pytorch core. With that, we can reduce the LoC needed in `pipeline.py`

File: torchtitan/train.py [hunk 7]
```diff
@@ -387,13 +387,10 @@ def forward_backward_step(
                     (labels, []) if self.pp_has_last_stage else (None, None)
                 )
                 if self.pp_has_first_stage:
-                    self.pp_schedule.step(
-                        inputs, target=targets, losses=losses, input_batch=inputs
-                    )
+                    # TODO: need to update kwargs for different models?
+                    self.pp_schedule.step(inputs, target=targets, losses=losses)
```
Comment (H-Huang) [inline torchtitan/train.py]: Llama3/4 [forward](https://github.com/pytorch/torchtitan/blob/01f4e50228fb4a304d17b5b5c5edfe39cc960000/torchtitan/models/llama3/model/model.py#L401) is slightly different from DeepSeekv3 [forward](https://github.com/pytorch/torchtitan/blob/0a6ab71b76d9831b81b27296522b230bed44ec3c/torchtitan/models/deepseek_v3/model/model.py#L361). It looks like the `input_batch` argument is needed for `flex_attn`.

What is the best way to go about this, should we create a new Trainer class for DeepSeekV3 that overrides `forward_backward_step`?

Comment (tianyu-l) [reply torchtitan/train.py]: let's reuse what llama3/4 has, which helps unblock document masking with block-causal attention masks.

File: torchtitan/models/deepseek_v3/infra/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+    Example:
+        generate_module_names_per_stage(2, 3, input_weight=2, output_weight=2)
+        treats embeddings as 2 layers and norm+output as 2 layers for distribution
+    """
+    if num_stages < 1:
+        raise ValueError("Number of stages must be at least 1")
+
+    if num_stages == 1:
+        # Single stage gets everything
+        layer_names = [f"layers.{i}" for i in range(num_layers)]
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/infra/pipeline.py:59]: did you take into consideration that the layers in dsv3 are not evenly distributed -- several dense layers, followed by MoE layers
https://github.com/pytorch/torchtitan/pull/1373/files#diff-ed005d894ae945a545c92c33136fba3bde35e70f1b7052f78242a1f69e862ab8R273

File: torchtitan/models/deepseek_v3/infra/pipeline.py (hunk 6 above)
Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/infra/pipeline.py:83]: makes sense, but can user still manually specify the split via `https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L313` or is it not encouraged any more?

Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/infra/pipeline.py:83]: also maybe we should try to upstream more functions in https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/pipeline.py

File: torchtitan/train.py (hunk 7 above)
Comment (H-Huang) [reply torchtitan/train.py]: Updated

File: torchtitan/models/deepseek_v3/infra/pipeline.py (hunk 8 above)
Comment (H-Huang) [reply torchtitan/models/deepseek_v3/infra/pipeline.py:59]: we don't take this into consideration yet 🫤.

**Comment by tianyu-l** [comment]
@H-Huang I'm so sorry -- I merged https://github.com/pytorch/torchtitan/pull/1373 and deleted the branch, which causes this PR to be closed.

**Comment by H-Huang** [comment]
@tianyu-l no worries haha, i will rebase

File: torchtitan/models/deepseek_v3/infra/pipeline.py (hunk 6 above)
Comment (H-Huang) [reply torchtitan/models/deepseek_v3/infra/pipeline.py:83]: > makes sense, but can user still manually specify the split via https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L313 or is it not encouraged any more?

I think we should move away from this

> also maybe we should try to upstream more functions in https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/pipeline.py

Yeah I agree, i think this can be cleaned up, let me think of a way

File: torchtitan/models/deepseek_v3/infra/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+                else:
+                    # No layers from this structure needed, set to empty structure
+                    if isinstance(module_value, nn.ModuleDict):
+                        setattr(model, module_name, nn.ModuleDict())
+                    elif isinstance(module_value, nn.ModuleList):
+                        setattr(model, module_name, nn.ModuleList())
+            # Handle simple module attributes (e.g., "linear", "norm")
+            elif module_name not in modules_to_keep:
+                # Replace with identity module instead of None
+                setattr(model, module_name, nn.Identity())
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/infra/pipeline.py]: what's the advantage of doing this instead of setting to `None`?
I'm worried about DCP loading / resharding, as multiple PP ranks will have the same fqns.

File: torchtitan/models/deepseek_v3/train_configs/debug_model.toml [hunk 10]
```diff
@@ -50,9 +50,11 @@ dataset = "c4_test"  # supported datasets: c4_test (2K), c4 (177M)
 data_parallel_replicate_degree = 1
 data_parallel_shard_degree = -1
 fsdp_reshard_after_forward = "default" # default / never / always
-tensor_parallel_degree = 2
+tensor_parallel_degree = 1
 enable_async_tensor_parallel = false
 expert_parallel_degree = 1
+pipeline_parallel_degree = 4
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/train_configs/debug_model.toml]: maybe still set this to 1 in debug_model

File: torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml [hunk 11]
```diff
@@ -51,6 +51,8 @@ fsdp_reshard_after_forward = "default" # default / never / always
 tensor_parallel_degree = 1
 enable_async_tensor_parallel = false
 expert_parallel_degree = 1
+pipeline_parallel_degree = 1
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml:54]: we can add this in the new 671B toml too

**Review by tianyu-l — COMMENTED** [review]
I feel it's fine to land this for now and the complete tests/benchmarks can come later. Could you show a screenshot of running FSDP+TP+PP together on 8 GPUs?

Besides, I had a question on DCP save / load / reshard. It'd be great if you can demonstrate an example of resharding during loading.

File: torchtitan/models/deepseek_v3/train_configs/debug_model.toml (hunk 10 above)
Comment (H-Huang) [reply torchtitan/models/deepseek_v3/train_configs/debug_model.toml]: yep will do

File: torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml (hunk 11 above)
Comment (H-Huang) [reply torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml:54]: done!

File: torchtitan/models/deepseek_v3/infra/pipeline.py (hunk 9 above)
Comment (H-Huang) [reply torchtitan/models/deepseek_v3/infra/pipeline.py]: that's a good point I didn't think of that. The main advantage was there isnt a need to change the model code and the tokens can just "pass through", but it seems that this is not robust enough so I'll change it back

**Review by tianyu-l — APPROVED** [review]
LGTM
